app/controllers/user_sessions_controller.rb

```diff
@@ -7,140 +7,162 @@ def new
-            # The identity is not associated with the current_user so lets
-            # associate the identity
-            @identity.save
-            @user = user
-            # log in them
-            @user_session = UserSession.create(@identity.user)
-            redirect_to root_url, notice: "Successfully linked to your account!"
-          end
+          redirect_to return_to, notice: "Successfully linked to your account!"
         end
       end
-    else
-      username = params[:user_session][:username] if params[:user_session]
-      u = User.find_by(username: username)
-      if u && u.password_checker != 0
-        n = u.password_checker
-        hash = { 1 => "Facebook", 2 => "Github", 3 => "Google", 4 => "Twitter"  }
-        s = "This account doesn't have a password set. It may be logged in with " + hash[n] + " account, or you can set a new password via Forget password feature"
-        flash[:error] = s
-        redirect_to '/'
-      else
-        params[:user_session][:username] = params[:openid] if params[:openid] # second runthrough must preserve username
-        @user = User.find_by(username: username)
-        # try finding by email, if that exists
-        if @user.nil? && !User.where(email: username).empty?
-          @user = User.find_by(email: username)
-          params[:user_session][:username] = @user.username
+    end
+  end
+
+  def handle_site_login_flow
+    username = params[:user_session][:username] if params[:user_session]
+    u = User.find_by(username: username)
+    if u && u.password_checker != 0
+      n = u.password_checker
+      hash = { 1 => "Facebook", 2 => "Github", 3 => "Google", 4 => "Twitter"  }
+      s = "This account doesn't have a password set. It may be logged in with " + hash[n] + " account, or you can set a new password via Forget password feature"
+      flash[:error] = s
+      redirect_to '/'
+      return
+    end
+
+    params[:user_session][:username] = params[:openid] if params[:openid] # second runthrough must preserve username
+    @user = User.find_by(username: username)
+    # try finding by email, if that exists
+    if @user.nil? && !User.where(email: username).empty?
+      @user = User.find_by(email: username)
+      params[:user_session][:username] = @user.username
+    end
+
+    if @user.nil?
+      flash[:warning] = "There is nobody in our system by that name, are you sure you have the right username?"
+      redirect_to '/login'
+      return
+    end
+
+    if @user&.drupal_user&.status == 5
+      flash[:error] = I18n.t('user_sessions_controller.user_has_been_moderated', username: @user.username).html_safe
+      redirect_to '/'
+      return
+    end
+
+    if @user&.drupal_user&.status.zero?
+      flash[:error] = I18n.t('user_sessions_controller.user_has_been_banned', username: @user.username).html_safe
+      redirect_to '/'
+      return
+    end
+
+    if @user&.drupal_user&.status == 1
+      # an existing Rails user
+      if @user
```

Comment:
@jywarren Is there any difference in `@user` and `@user.nil` in this case ? If not, then we can remove if condition because we already have check for that in above code.

Reply:
I think you're right, here. Let's just be very careful and make changes incrementally since this is a pretty complex piece of code. It'll be easier to track any bugs we may introduce if we change fewer things at a time. I know this is not as satisfying as more thorough refactor here, but our login code is relatively critical, so we have to be careful, and not every thing can be protected by tests. Thanks!